cirq/value/value_equality.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+def _value_equality_eq(self, other):
+    cls_self = self._value_equality_values_cls_()
+    if not isinstance(other, cls_self):
+        return NotImplemented
+    cls_other = other._value_equality_values_cls_()
+    if cls_self != cls_other:
+        return False
+    return self._value_equality_values_() == other._value_equality_values_()
+
+
+def _value_equality_ne(self, other):
+    return not self == other
+
+
+def _value_equality_hash(self):
+    return hash((self._value_equality_values_cls_(),
+                 self._value_equality_values_()))
+
+
+def value_equality(*args,
+                   unhashable: bool = False,
+                   distinct_child_types: bool = False):
```

Comment:
can we add a type annotation for the return value?

Reply:
Done. I avoided it before because it required declaring overloads (the return type is `Union[Callable[[type], type], type]` and when you return the lambda it has apparently-incompatible type `Callable[[type], Union[Callable[[type], type], type]]`, so mypy was angry without the overloads).